### PR TITLE
[Merged by Bors] - feat: port Algebra.Order.Sub.Defs

### PR DESCRIPTION
**Pull request by rosborn**
Based on '71ca477041bcd6d7c745fe555dc49735c12944b7'

### CODE REVIEW

File: Mathlib/Algebra/Order/Sub/Defs.lean [hunk 1]
```diff
@@ -0,0 +1,452 @@
...
+import Mathlib.Order.Lattice
+
+/-!
+# Ordered Subtraction
+
+This file proves lemmas relating (truncated) subtraction with an order. We provide a class
+`HasOrderedSub` stating that `a - b ≤ c ↔ a ≤ c + b`.
+
+The subtraction discussed here could both be normal subtraction in an additive group or truncated
+subtraction on a canonically ordered monoid (`ℕ`, `Multiset`, `PartEnat`, `Ennreal`, ...)
```
Comment (semorrison) [inline Mathlib/Algebra/Order/Sub/Defs.lean:18]:
```suggestion
subtraction on a canonically ordered monoid (`ℕ`, `Multiset`, `PartENat`, `ENnReal`, ...)
```
Not entirely sure about `ENnReal`, but I think that's what the current rule says!

File: Mathlib/Algebra/Order/Sub/Defs.lean [hunk 2]
```diff
@@ -0,0 +1,452 @@
...
+
+The results in this file are ordered by the type-class assumption needed to prove it.
+This means that similar results might not be close to each other. Furthermore, we don't prove
+implications if a bi-implication can be proven under the same assumptions.
+
+Lemmas using this class are named using `tsub` instead of `sub` (short for "truncated subtraction").
+This is to avoid naming conflicts with similar lemmas about ordered groups.
+
+We provide a second version of most results that require `[ContravariantClass α α (+) (≤)]`. In the
+second version we replace this type-class assumption by explicit `AddLeCancellable` assumptions.
```
Comment (semorrison) [inline Mathlib/Algebra/Order/Sub/Defs.lean:35]:
```suggestion
second version we replace this type-class assumption by explicit `AddLECancellable` assumptions.
```

File: Mathlib/Algebra/Order/Sub/Defs.lean [hunk 3]
```diff
@@ -0,0 +1,452 @@
...
+#align add_tsub_add_le_tsub_right add_tsub_add_le_tsub_right
+
+end Cov
+
+/-! #### Lemmas that assume that an element is `AddLeCancellable` -/
+
+
+namespace AddLeCancellable
+
+protected theorem le_add_tsub_swap (hb : AddLeCancellable b) : a ≤ b + a - b :=
```
Comment (semorrison) [inline Mathlib/Algebra/Order/Sub/Defs.lean:198]:
```suggestion
protected theorem le_add_tsub_swap (hb : AddLECancellable b) : a ≤ b + a - b :=
```
etc

File: Mathlib/Algebra/Order/Sub/Defs.lean [hunk 4]
```diff
@@ -0,0 +1,452 @@
...
+  rw [add_comm]
+  apply tsub_add_eq_tsub_tsub
+#align tsub_add_eq_tsub_tsub_swap tsub_add_eq_tsub_tsub_swap
+
+theorem tsub_right_comm : a - b - c = a - c - b := by
+  rw [←tsub_add_eq_tsub_tsub, tsub_add_eq_tsub_tsub_swap]
+
+#align tsub_right_comm tsub_right_comm
+
+/-! ### Lemmas that assume that an element is `add_le_cancellable`. -/
```
Comment (semorrison) [inline Mathlib/Algebra/Order/Sub/Defs.lean:280]:
```suggestion
/-! ### Lemmas that assume that an element is `AddLECancellable`. -/
```

File: Mathlib/Algebra/Order/Sub/Defs.lean (hunk 2 above)
Comment (rosborn) [reply Mathlib/Algebra/Order/Sub/Defs.lean:35]: This disagrees with Algebra.Order.Monoid.Lemmas. Should that file also be changed?

**Review by Ruben-VandeVelde — APPROVED** [review]
Good to go, I think

File: Mathlib/Algebra/Order/Sub/Defs.lean (hunk 4 above)
Comment (semorrison) [reply Mathlib/Algebra/Order/Sub/Defs.lean:280]: Looks like this didn't actually happen?

**Comment by semorrison** [comment]
bors merge

**Pull request closed without merging**